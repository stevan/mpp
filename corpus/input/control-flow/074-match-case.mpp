# Basic match/case structure

match ($value) {
    case (1) {
        print("one");
    }
    case (2) {
        print("two");
    }
    case (3) {
        print("three");
    }
    else {
        print("other");
    }
}

# Match with multiple values
match ($day) {
    case (1) {
        print("Monday");
    }
    case (2) {
        print("Tuesday");
    }
    else {
        print("Weekend");
    }
}

# Match with expressions
match ($score) {
    case ($_ > 90) {
        print("A");
    }
    case ($_ > 80) {
        print("B");
    }
    else {
        print("F");
    }
}

# Match in expression context
my $result = match ($code) {
    case (200) { "success" }
    case (404) { "not found" }
    else { "unknown" }
};