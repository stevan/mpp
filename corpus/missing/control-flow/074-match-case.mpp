# Basic match/case structure (experimental)
use feature 'switch';

match ($value) {
    case (1) {
        print("one");
    }
    case (2) {
        print("two");
    }
    case (3) {
        print("three");
    }
    else {
        print("other");
    }
}

# Match with patterns
match ($string) {
    case (/^\d+$/) {
        print("all digits");
    }
    case (/^[a-z]+$/) {
        print("all lowercase");
    }
    case (/^[A-Z]+$/) {
        print("all uppercase");
    }
    else {
        print("mixed");
    }
}

# Match with multiple values
match ($day) {
    case (1..5) {
        print("weekday");
    }
    case (6, 7) {
        print("weekend");
    }
}

# Match with conditions
match ($score) {
    case ($_ > 90) {
        print("A");
    }
    case ($_ > 80) {
        print("B");
    }
    case ($_ > 70) {
        print("C");
    }
    else {
        print("F");
    }
}

# Nested match
match ($type) {
    case ("number") {
        match ($value) {
            case ($_ < 0) {
                print("negative");
            }
            case ($_ > 0) {
                print("positive");
            }
            else {
                print("zero");
            }
        }
    }
    case ("string") {
        process_string($value);
    }
}

# Match in expression context
my $result = match ($code) {
    case (200) { "success" }
    case (404) { "not found" }
    case (500) { "server error" }
    else { "unknown" }
};

# Match with complex patterns
match ($input) {
    case ([1, 2, 3]) {
        handle_low();
    }
    case ([4, 5, 6]) {
        handle_mid();
    }
    case ([7, 8, 9]) {
        handle_high();
    }
    case (0) {
        handle_zero();
    }
}