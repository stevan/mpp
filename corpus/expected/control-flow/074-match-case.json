[
  {
    "type": "Match",
    "expression": {
      "type": "Variable",
      "name": "$value"
    },
    "caseClauses": [
      {
        "pattern": {
          "type": "Number",
          "value": "1"
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"one\""
              }
            ]
          }
        ]
      },
      {
        "pattern": {
          "type": "Number",
          "value": "2"
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"two\""
              }
            ]
          }
        ]
      },
      {
        "pattern": {
          "type": "Number",
          "value": "3"
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"three\""
              }
            ]
          }
        ]
      }
    ],
    "elseBlock": [
      {
        "type": "Call",
        "name": "print",
        "arguments": [
          {
            "type": "String",
            "value": "\"other\""
          }
        ]
      }
    ]
  },
  {
    "type": "Match",
    "expression": {
      "type": "Variable",
      "name": "$day"
    },
    "caseClauses": [
      {
        "pattern": {
          "type": "Number",
          "value": "1"
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"Monday\""
              }
            ]
          }
        ]
      },
      {
        "pattern": {
          "type": "Number",
          "value": "2"
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"Tuesday\""
              }
            ]
          }
        ]
      }
    ],
    "elseBlock": [
      {
        "type": "Call",
        "name": "print",
        "arguments": [
          {
            "type": "String",
            "value": "\"Weekend\""
          }
        ]
      }
    ]
  },
  {
    "type": "Match",
    "expression": {
      "type": "Variable",
      "name": "$score"
    },
    "caseClauses": [
      {
        "pattern": {
          "type": "BinaryOp",
          "operator": ">",
          "left": {
            "type": "Variable",
            "name": "$_"
          },
          "right": {
            "type": "Number",
            "value": "90"
          }
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"A\""
              }
            ]
          }
        ]
      },
      {
        "pattern": {
          "type": "BinaryOp",
          "operator": ">",
          "left": {
            "type": "Variable",
            "name": "$_"
          },
          "right": {
            "type": "Number",
            "value": "80"
          }
        },
        "block": [
          {
            "type": "Call",
            "name": "print",
            "arguments": [
              {
                "type": "String",
                "value": "\"B\""
              }
            ]
          }
        ]
      }
    ],
    "elseBlock": [
      {
        "type": "Call",
        "name": "print",
        "arguments": [
          {
            "type": "String",
            "value": "\"F\""
          }
        ]
      }
    ]
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$result"
    },
    "initializer": {
      "type": "Match",
      "expression": {
        "type": "Variable",
        "name": "$code"
      },
      "caseClauses": [
        {
          "pattern": {
            "type": "Number",
            "value": "200"
          },
          "block": [
            {
              "type": "String",
              "value": "\"success\""
            }
          ]
        },
        {
          "pattern": {
            "type": "Number",
            "value": "404"
          },
          "block": [
            {
              "type": "String",
              "value": "\"not found\""
            }
          ]
        }
      ],
      "elseBlock": [
        {
          "type": "String",
          "value": "\"unknown\""
        }
      ]
    }
  }
]